Step2 리뷰 요청 드립니다.

itdar opened the pull request:
안녕하세요 선동님!~
볼링게임 규칙을 제대로 알게된건 처음인 것 같은데
생각보다 점수계산이 골치 아프겠더라구요 ㅎㅎ;

Controller - Game - Frames - Frame - Pins/Scores 구조로 되어있습니다.
Controller/Game 은 관점에 따라서 둘 다 클라이언트 역할이 될 수 있을 것 같고..
Frames 는 Frame 자체로 연결리스트 구조와 ArrayList 를 쓴 것 두개를 유지하다가 연결리스트 형태만 남기고 제거했습니다.
볼링을 안다는 가정 하에 프레임 자체로 연결되는 구조가 이해하기 좀 더 쉬워 좋아보여서 굳이 배열을 쓰지 않게 되더라구요.
Pins 는 볼링핀 전체이고
Scores 는 1/2차 시도에 넘어뜨린 점수 입니다.

리뷰 감사합니다!~

itdar committed:
refactor: 선동님 리뷰사항 적용 커밋 #2

- 간소화 할 수 있는 로직 정리
- EOL warning 없도록 IDE 설정
- 예외처리 문자열 상수 처리

itdar committed:
feat: bowling step2 점수판 그리기 최초 커밋

- Client -> Contnroller -> Domain 구조
- 도메인에는,
  - Frames (해당 참가자가 갖는 프레임 당 점수 정보들 (프레임 연결리스트의 헤드 및 현재 노드주소 등)
  - Frame 각 프레임의 첫번째/두번째 점수 등
  - Score 점수
  - Pins 볼링핀
  - Strike, HitState -> 볼링 규칙보다가 그냥 미리 만들어둔 것들
- ui in/out 각각의 클래스

itdar committed:
feat: 유저 입력 후 빈 프레임들 출력 포맷까지 맞춤

itdar committed:
feat: 볼링 게임 점수계산 기능기능 구현 완료

itdar committed:
refactor: 리팩토링

- Controller 와 Frames 사이에 Game 클래스 추가
- Frames 연결리스트 외에 배열로 추가 관리하던 것 제거
- Frame 에 1,2차 score 패키징

itdar committed:
feat: 마지막프레임(10번째)에서 스트라이크/스패어 처리 시 연장게임 기능 추가

seondongpyo commented on src/main/java/bowling/MainApp.java:
```diff
@@ -0,0 +1,33 @@
+        OutputCui outputCui = new OutputCui();
+        BowlingGameController bowlingGame = new BowlingGameController();
+
+        final String name = inputCui.inputPlayerNames();
```
`InputCui`의 `inputParticipantsImpl()`를 직접 호출해도 되지 않을까요?
`inputPlayerNames()`라는 메서드로 감싼 의도가 따로 있으실까요?

itdar replied:
옙옙 처음에 상위 메서드에서 횟수 제한을 할까해서 넣었다가 지웠는데
남아있었네요 ㅎㅎ; 통합해서 한겹으로 남겨둬보겠습니다!~

seondongpyo commented on src/main/java/bowling/MainApp.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    private static void play(InputCui inputCui, OutputCui outputCui, Game game) {
+        int hitCount;
```
해당 지역 변수는 `while`문 안에서 선언해도 괜찮지 않을까 생각합니다.

itdar replied:
오 말씀해주신 것을 보고 찾아보니 여러번 할당해서 사용하는 객체들이 아니면
메모리 사용에 차이가 없어서 스코프를 고려해서 의미적으로 맞게 쓰는게 낫다 같은데 이렇게 보시나요??

seondongpyo replied:
넵 아예 인라인으로 작성하여 지역 변수를 사용하지 않는 방법도 있을 것 같아요.
```
game.throwBall(inputCui.inputHitCount(game.frameNumber()));
```

seondongpyo commented on src/main/java/bowling/domain/FinalFrame.java:
```diff
@@ -0,0 +1,22 @@
+    public void shot(int hitCount) {
+        super.shot(hitCount);
+
+        if (scores.isStrike()) {
+            createNext().createNext();
+            return;
+        }
+
+        if (scores.isSpare()) {
+            createNext();
+        }
```
마지막 프레임에서 스트라이크 또는 스페어를 기록할 경우, 
프레임이 아닌 투구 기회가 추가되어야 하지 않을까요? 🤔

itdar replied:
아흑 그러네요 잘못 생각했네요 ㅋㅋㅋㅋ... 수정해보겠습니다!~

seondongpyo commented on src/main/java/bowling/domain/Frames.java:
```diff
@@ -0,0 +1,51 @@
+
+public class Frames {
+
+    protected static final int BOWLING_NORMAL_FRAMES = 9;
+    private Frame head;
```
다른 클래스에서처럼 상수와 인스턴스 변수의 구분을 위해 공백 라인을 추가하면 좋을 것 같습니다.
```suggestion
    protected static final int BOWLING_NORMAL_FRAMES = 9;
    
    private Frame head;
```

itdar replied:
엇 옙옙 수정해보겠습니다~!

seondongpyo commented on src/main/java/bowling/domain/HitState.java:
```diff
@@ -0,0 +1,24 @@
+import static java.lang.Boolean.FALSE;
+import static java.lang.Boolean.TRUE;
+
+public enum HitState {
```
현재는 테스트 코드에서만 사용되는 것으로 보이는데, 다음 단계를 위해 미리 추가해놓으신 거라고 봐도 될까요?

itdar replied:
옙 맞습니다 ㅎㅎ; 처음에 볼링규칙이랑 점수관련 단어 보다가 
Strike 랑 아래에 말씀하신 State 이넘을 만들어뒀었는데 지울까 하다가 나중에 규칙이라도 볼겸 아니면 기능으로 쓰려나 해서 냅뒀네요;

seondongpyo commented on src/main/java/bowling/domain/Pins.java:
```diff
@@ -0,0 +1,47 @@
+
+    public static final int DEFAULT_PIN = 10;
+
+    public int pins;
```
접근 제어자를 `private`으로 선언해야 하지 않을까요? 
외부에서 볼링 핀의 숫자가 변경되어 의도치 않은 경우가 발생할 여지가 있어 보입니다.

itdar replied:
어이쿠 정신이 없었네요 ㅎㅎ;;

seondongpyo commented on src/main/java/bowling/domain/Scores.java:
```diff
@@ -0,0 +1,63 @@
+        secondScore = new Score(hitCount);
+    }
+
+    public boolean isStrike() {
+        return firstScore.get() == 10;
+    }
+
+    public boolean isSpare() {
+        if (firstScore == null || secondScore == null)
+            return false;
+
+        return firstScore.get() + secondScore.get() == 10;
+    }
```
숫자 리터럴 `10`을 기존에 선언된 상수를 활용하거나 새로운 상수로 선언하여 의미를 부여해보면 어떨까요?

itdar replied:
옙 매직넘버를 없애보겠습니다!~

seondongpyo commented on src/main/java/bowling/domain/Strike.java:
```diff
@@ -0,0 +1,46 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public enum Strike {
```
해당 enum도 `HitState`와 마찬가지로 어떤 의도로 작성하신 건지 궁금합니다.

itdar replied:
옙 위와 같이 규칙 보면서 메모 겸 미리 만들어뒀는데 아직 쓸데가 없어서..; ㅋㅋ
점수 계산 떄 쓰이거나 아니면 없애보겠습니다!~

seondongpyo commented on src/main/java/bowling/ui/InputCui.java:
```diff
@@ -0,0 +1,40 @@
+
+import java.util.Scanner;
+
+public class InputCui {
+    private static final int NAME_LENGTH = 3;
+
+    private Scanner scanner = new Scanner(System.in);
+
+    public InputCui() { }
```
`View`와 같은 클래스는 유틸리티 클래스로 만들어보는 건 어떻게 생각하시나요?
`View`에서 상태를 가질 필요가 없다면, 상수와 정적 메서드만을 갖도록 하여
생성자를 통해 인스턴스를 생성하지 않아도 사용할 수 있게 해보면 좋을 것 같습니다.

itdar replied:
옙 좋은 것 같습니다! `View` 내의 Input / Output 클래스는 유틸리티 클래스로 바꿔보겠습니다!!

seondongpyo commented on src/main/java/bowling/ui/InputCui.java:
```diff
@@ -0,0 +1,40 @@
+        out.printf("플레이어 이름은(%d english letters)?: ", NAME_LENGTH);
+        String nameInitial = scanner.nextLine();
+
+        if (!isValidNames(nameInitial)) {
+            return inputParticipantsImpl();
+        }
```
입력한 이름이 왜 유효하지 않은지를 메시지로 출력한다면, 사용자 편의성이 더 높아질 것 같아요.

itdar replied:
넵 출력할 문자를 추가해보겠습니다!~

seondongpyo commented on src/main/java/bowling/ui/OutputCui.java:
```diff
@@ -0,0 +1,73 @@
+import bowling.domain.Score;
+import java.util.Objects;
+
+public class OutputCui {
```
하드 코딩된 숫자 및 문자열 리터럴들을 상수로 선언해보는 건 어떨까요?

itdar replied:
엇 그쵸 이건 나중에 점수 계산할 때 이넘과 같이 쓸 수 있지 않을까 하긴 했었는데요 ㅎㅎ;
일단 상수로 바꿔보겠습니다~!

seondongpyo commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -77,15 +77,17 @@ public String toString() {
 
         delete();
+
+        return DeleteHistory.of(this);
```
유효하지 않은 메서드 호출로 보이네요. 컴파일 에러를 확인해주시면 좋을 것 같습니다.

itdar replied:
옙 이건 1단계 이후에 수정한 내용을 같이 커밋하지 않았었네요; 다른 것 고치기 전에 커밋해두었습니다!~

seondongpyo commented on src/test/java/bowling/domain/FramesTest.java:
```diff
@@ -0,0 +1,130 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+class FramesTest {
```
위에서 말씀드렸지만, 요구사항 및 볼링 규칙을 다시 확인하시고 테스트를 작성해보시면 좋을 것 같습니다.

seondongpyo commented on src/test/java/bowling/domain/PinsTest.java:
```diff
@@ -0,0 +1,37 @@
+    @Test
+    void 남아있는_핀_이상의_숫자를_쳤다고_들어오면_예외발생한다() {
+        // given
+        final int hitCount = 11;
```
볼링 핀 개수의 경계값인 `11`을 이용하여 테스트를 잘 작성해주셨네요 👍
또 다른 경계값인 `-1`에 대한 경우도 테스트해보면 어떨까요?

itdar replied:
옙옙 반대편 엣지도 추가해보겠습니다!~

seondongpyo commented on src/test/java/bowling/domain/ScoresTest.java:
```diff
@@ -0,0 +1,38 @@
+        assertThat(scores.second()).isNull();
+    }
+
+    @DisplayName("첫번째 두번째 볼을 굴려 맞춘 개수가 맞다.")
+    @Test
+    void hitTest() {
+        Scores scores = new Scores();
+
+        scores.hit(4);
+        scores.hit(6);
+    }
```
- 현재 해당 테스트에 검증 로직이 빠져 있는데, 추가해보시면 좋을 것 같습니다.
- 테스트 설명을 조금 더 명확하게 작성해보면 어떨까요? '맞춘 개수'가 무엇과 '맞다'는 걸 의도하셨을까요?

itdar replied:
아.. 1/2차시기에 굴려서 맞춘 개수(점수) 가 제대로 저장되었다는 의미였는데 뒤에 검증이 빠졌네요; ㅎㅎ
추가해보겠습니다!!

seondongpyo reviewed (changes requested):
볼링 점수판 그리기 미션도 잘 구현해주셨네요 👍
구현하신 내용과 관련하여 몇 가지 코멘트를 드렸는데,
특히 이번 미션의 요구사항과 볼링 규칙을 다시 확인해주시면 좋을 것 같아요.
조금만 더 힘내주시고 다시 리뷰 요청 부탁드릴게요 🔥

itdar committed:
fix: 1단계 수정했던 내용 커밋

itdar committed:
refactor: 선동님 리뷰사항 수정 커밋

- 빠진 테스트 추
- 매직넘버, 리터럴 상수로 변경
- 유틸성 클래스로 변경 (in/out)
- 10프레임에서 프레임 추가 없이 추가 볼 구조로 수정
- 가시성 변경

seondongpyo commented on src/main/java/bowling/ui/InputCui.java:
```diff
@@ -0,0 +1,37 @@
+        String nameInitial = scanner.nextLine();
+
+        if (!isValidNames(nameInitial)) {
+            out.printf("플레이어의 이름 길이는 %d 의 문자열만 가능합니다.", NAME_LENGTH);
```
마지막에 줄바꿈 문자를 추가해주면 가독성이 더 좋아질 것 같습니다 🙂

![image](https://user-images.githubusercontent.com/64854054/169925239-4d459ef7-9f9a-4f47-96bc-b64da7a6f0be.png)

itdar replied:
엇 빠뜨렸네요 추가해두겠습니다~!

seondongpyo commented on src/main/java/bowling/ui/OutputCui.java:
```diff
@@ -0,0 +1,109 @@
+        out.print(GAP + SCORE_SEPARATOR);
+    }
+
+    private static String drawFirstScore(Score score) {
```
첫 번째 투구 결과가 거터인 경우도 처리해보면 어떨까요?

![image](https://user-images.githubusercontent.com/64854054/169924941-2bace036-e2f5-4b9e-b549-b9e63c7894b4.png)

itdar replied:
아.. 거터 다시 봐서 넣어보겠습니다!!

seondongpyo commented on src/test/java/bowling/domain/FramesTest.java:
```diff
@@ -0,0 +1,130 @@
+        assertThat(frames).isNotNull();
+    }
+
+    @DisplayName("프레임들 객체에서 볼링게임 기본프레임 10개를 생성한다.")
```
```suggestion
    @DisplayName("프레임들 객체에서 볼링게임 기본프레임 9개를 생성한다.")
```

itdar replied:
갯수에 맞게 수정해보겠습니다!!

seondongpyo commented on src/test/java/bowling/domain/PinsTest.java:
```diff
@@ -0,0 +1,49 @@
+        }).isInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    void 음수의_핀을_쳤다고_들어오면_예외발생한다() {
+        // given
+        final int hitCount = -1;
+        Pins pins = new Pins();
+        assertThat(pins.standing()).isEqualTo(START_PIN_COUNT);
+
+        assertThatThrownBy(() -> {
+            pins.hit(hitCount);
+        }).isInstanceOf(IllegalArgumentException.class);
+    }
```
👍👍 
여기도 `@ParameterizedTest`, `@ValueSource` 등을 활용하여 중복 코드를 제거해보면 좋을 것 같아요 😃

itdar replied:
안쓰다보니 문법을 잊었네요 ㅎㅎ;; 다시 봐서 써보겠습니다!~

seondongpyo reviewed (approved):
피드백을 잘 반영해주셨네요 👍
추가로 몇 가지 코멘트를 드렸는데, 다음 단계에서 확인해주시면 좋을 것 같습니다.
이번 단계는 여기서 merge 하도록 할게요.
다음 단계도 힘내시길 바랍니다 🔥

itdar commented:
> 피드백을 잘 반영해주셨네요 👍 추가로 몇 가지 코멘트를 드렸는데, 다음 단계에서 확인해주시면 좋을 것 같습니다. 이번 단계는 여기서 merge 하도록 할게요. 다음 단계도 힘내시길 바랍니다 🔥

리뷰 열심히 잘 해주셔서 감사합니다~~!!
다음 단계 시작 전에 수정커밋 해두고 시작하겠습니다
감사합니다!!
